[WIP] Revise deployment overview for publish/deploy workflow and integration extensibility (#4812)
dotnet/docs-aspire: docs/deployment/overview.md

Reviewer: Replace all occurrences of ".NET Aspire" with "Aspire".

---
title: .NET Aspire deployments

> description: Understand how to publish and deploy .NET Aspire applications using the Aspire CLI, hosting integrations, and related tooling.

Reviewer 2: revert this

- `DeployingCallbackAnnotation` can be added to a resource to support `aspire deploy`

This design enables hybrid and heterogeneous deployments, where different services within the same app can be deployed to different targets (cloud, edge, local).

### Compute environments

A **compute environment** is a core deployment concept in .NET Aspire that represents a target platform where your application resources will be deployed. Compute environments define how resources should be transformed and what deployment artifacts should be generated.

**Compute resources** are the runnable parts of your application, such as .NET projects, containers, and executables that need to be deployed to a compute environment.

An ACA environment is like a secure container app cluster. Most applications only need one such cluster, since it provides internal networking, scaling, and service discovery. Multiple environments are only required when you need regional separation, strict isolation, or organizational boundaries. For most apps, those cases don't apply. Most often, the following code is sufficient:

```csharp
builder.AddAzureContainerAppEnvironment("env");
```

> When you add a compute environment, Aspire applies the correct publishing behavior to all compatible compute resources in your app model—no extra configuration needed.

Reviewer 2: Use DockerCompose and Kuberentes asn example of compute environments.

> Deploy support is integration-specific. Absence of deploy support means you use the published artifacts with external tooling.

## Typical workflows

### 1. Generate artifacts (any integration)

```bash
aspire publish -o artifacts/
```

Review the contents of `artifacts/` (for example: Docker Compose files, Kubernetes manifests, Azure specification documents, etc.).

### 2. Run locally (Docker example)

```bash
# Provide or export required environment variables, then:
docker compose -f artifacts/docker-compose.yml up --build
```

Missing variables like `PG_PASSWORD` must be set in the shell, an `.env` file, or injected by your chosen runner.

### 3. CI/CD pipeline pattern

1. Run `aspire publish -o artifacts/` in a build job.
2. Archive or inspect artifacts (linting, security scanning, image signing).
3. Inject parameter values (environment variables, secrets store, variable groups).
4. Execute deployment:
   - Docker: `docker compose up -d`
   - Kubernetes: `kubectl apply -f artifacts/`
   - Azure: Use Azure CLI or deployment scripts with the published artifacts
   - Custom: your own script or orchestrator.

> Alternative for Azure: If using integrations with deploy support, you can use `aspire deploy` instead of publishing and deploying separately.

Reviewer 2: Remove this.

> Alternative for Azure: If using integrations with deploy support, you can use `aspire deploy` instead of publishing and deploying separately.

Reviewer 2: Remove this one.

### 4. Using `aspire deploy`

If an integration supports deployment, you can run:

```bash
aspire deploy
```

This resolves parameters and applies deployment changes for integrations that support deployment.

## Extensibility

The system is designed so you can author new integrations that participate in:

- Publish phase: generating artifacts for a target platform (for example, a service mesh config, an infrastructure-as-code definition, or a proprietary orchestrator spec).
- Deploy phase: applying resources to a new deployment target (for example, a proprietary app hosting environment).

A custom integration can:

- Transform application resources into platform-specific configurations.
- Handle parameter resolution and secret management.
- Implement deployment workflows for custom platforms.

> Because published assets are parameterized, you can integrate custom secret and configuration resolution without modifying the core publishing logic.

Reviewer 2: Tihs section should be focused around adding `PublishingCallbackAnnotation` and `DeployingCallbackAnnotation` to resources which give control over defining what happens during aspire publish or aspire  deploy. This is still in preview and we expect changes to the design.

Use the example and text from https://github.com/dotnet/docs-aspire/blob/main/docs/whats-new/dotnet-aspire-9.4.md#aspire-deploy in this section.

---
title: Aspire publishing and deployment
description: Learn how Aspire solutions are published and deployed.
ms.topic: overview
ms.date: 09/26/2025
---

# Aspire publishing and deployment overview

Aspire separates the act of _producing deployment assets_ from _executing a deployment_. The Aspire CLI (`aspire`) provides two high‑level entrypoints:

> - `aspire publish`: Generates intermediate, parameterized assets for one or more hosting integrations.

Reviewer 3: ```suggestion
- `aspire publish`: Generates intermediate, parameterized assets for one or more hosting integrations that implement publish semantics.
```

      HTTP_PORTS: "8001"
      ConnectionStrings__db: "Host=pg;Port=5432;Username=postgres;Password=${PG_PASSWORD};Database=db"
    ports:
      - "8002:8001"
      - "8004:8003"
    depends_on:
      pg:
        condition: "service_started"
    networks:
      - "aspire"
```

Key points:

- `${PG_PASSWORD}` and `${DBSETUP_IMAGE}` (and similar) are *placeholders* in the published asset.
- They are not resolved during `aspire publish`.
- A deployment engine (which could be `aspire deploy`, `docker compose` plus a script exporting variables, CI/CD variable injection, etc.) supplies their values later.
- This keeps secrets and environment-specific values decoupled from generated structure.

Different integrations may use different placeholder conventions (environment variables, tokens, or parameter metadata), but the principle remains: publish preserves *shape*, deploy injects *values*.

## Publisher model and compute environments

Aspire uses a flexible publisher model that distributes publishing behavior across your application graph. Resources support publishing and deployment through annotations:

- `aspire publish`
- `aspire deploy`

> This design enables hybrid and heterogeneous deployments, where different services within the same app can be deployed to different targets (cloud, edge, local).

Reviewer 3: I think this statement hides some of the complexity here. We only support one of these annotatations per-resource at the moment so a hosting integration has to implement multiple resources to do the right thing. I think phrasing like:

> This annotation-based approach means that different hosting integrations within the same app can implement different publish and deployment behavior, supporting deployments to different targets.

When you add a compute environment like Docker Compose or Kubernetes, Aspire applies the correct publishing behavior to all compatible compute resources in your app model—no extra configuration needed.

### Multiple environments require disambiguation

If you add multiple compute environments, Aspire needs to know which resource goes where. Compute environments apply their transformations to all applicable compute resources (projects, containers, executables). If more than one environment matches a given resource, Aspire throws an ambiguous environment exception at publish time.

You can resolve this by using `WithComputeEnvironment(...)`:

```csharp
var k8s = builder.AddKubernetesEnvironment("k8s-env");
var compose = builder.AddDockerComposeEnvironment("docker-env");

builder.AddProject<Projects.Frontend>("frontend")
    .WithComputeEnvironment(k8s);

builder.AddProject<Projects.Backend>("backend")
    .WithComputeEnvironment(compose);
```

This example shows how you could explicitly map services to different compute environments. For example, a frontend in Kubernetes and a backend in Docker Compose.

## Hosting integration support matrix

> | Integration Package | Target | Publish | Deploy | Notes |
> |---------------------|--------|---------|--------|-------|

Reviewer: ```suggestion
| Integration package | Target | Publish | Deploy | Notes |
|---------------------|--------|---------|--------|-------|
```

| `Aspire.Hosting.Docker` | Docker / Docker Compose | ✅ Yes | ❌ No | Use generated Compose with your own scripts or tooling. |
| `Aspire.Hosting.Kubernetes` | Kubernetes | ✅ Yes | ❌ No | Apply with `kubectl`, GitOps, or other controllers. |
| `Aspire.Hosting.Azure.AppContainers` | Azure Container Apps | ✅ Yes | ✅ Yes (Preview) | Deploy capability is in Preview and may change. |
| `Aspire.Hosting.Azure.AppService` | Azure App Service | ✅ Yes | ✅ Yes (Preview) | Deploy capability is in Preview and may change. |

> [!TIP]
> Deploy support is integration-specific. Absence of deploy support means you use the published artifacts with external tooling.

## Typical workflows

### 1. Generate artifacts (any integration)

> ```bash

Reviewer: ```suggestion
```Aspire
```

aspire publish -o artifacts/
```

> Review the contents of `artifacts/` (for example: Docker Compose files, Kubernetes manifests, Azure specification documents, etc.).

Reviewer: ```suggestion
Review the contents of _artifacts/_ (for example: Docker Compose files, Kubernetes manifests, Azure specification documents, etc.).
```

### 2. Run locally (Docker example)

```bash
# Provide or export required environment variables, then:
docker compose -f artifacts/docker-compose.yml up --build
```

Missing variables like `PG_PASSWORD` must be set in the shell, an `.env` file, or injected by your chosen runner.

### 3. Using `aspire deploy`

If an integration supports deployment, you can run:

> ```bash

Reviewer: ```suggestion
```Aspire
```

                    Required = true,
                    Placeholder = "dev, staging, prod"
                },
                cancellationToken: ct);

            // Custom deployment logic here
            var reporter = ctx.ActivityReporter;
            await using (var deployStep = await reporter.CreateStepAsync(
                $"Deploying data seed job to {envResult.Value}", ct))
            {
                // Simulate deployment work
                await Task.Delay(2000, ct);
                await deployStep.SucceedAsync("Data seed job deployed successfully", ct);
            }
        }));

        return resourceBuilder;
    }
}
```

This custom deployment logic integrates seamlessly with the `aspire deploy` command, providing interactive prompts and progress reporting.

## Diagnostics & auditing

Publishing gives you an immutable snapshot of intended structure before secrets appear. You can:

- Diff published outputs between commits.
- Scan for disallowed images or configuration.
- Preserve a record for compliance, then separately record the resolved set applied at deployment time.

## Additional tools

### Azure Developer CLI (`azd`)

> [Azure Developer CLI (azd)](https://learn.microsoft.com/azure/developer/azure-developer-cli/) has first-class support for deploying Aspire projects. It can provision infrastructure, manage environments, and coordinate secret/value injection. You can incorporate Aspire publish artifacts into `azd` workflows or use the Azure integration (preview) directly.

Reviewer: ```suggestion
[Azure Developer CLI (azd)](/azure/developer/azure-developer-cli/) has first-class support for deploying Aspire projects. It can provision infrastructure, manage environments, and coordinate secret/value injection. You can incorporate Aspire publish artifacts into `azd` workflows or use the Azure integration (preview) directly.
```

> ### Aspir8 (Kubernetes YAML generation)
> 
> <a href="https://prom3theu5.github.io/aspirational">Aspir8 (Aspirate)</a> is a community tool that can transform an Aspire application model into Kubernetes manifests. While the Kubernetes hosting integration covers publish generation, Aspir8 may offer additional transforms or workflow preferences.
> 

Reviewer: I think this doesn't exist anymore.

```suggestion
```

> ## Legacy deployment manifest (footnote)

Reviewer: Are we ok calling the manifest "legacy"?

Earlier workflows emphasized a single "deployment manifest" generated from specialized AppHost targets. The modern approach centers on `aspire publish` + integration extensibility. The legacy manifest format is **not being evolved further**, but you can still generate it for inspection or debugging:

> ```bash

Reviewer: ```suggestion
```Aspire
```

> aspire publish --publisher manifest -o diagnostics/

Reviewer: This isn't how it was originally done, it was with a `dotnet` command...does this work?

```

This:

- Produces a manifest snapshot useful for understanding resource graphs or troubleshooting.
- Should not be treated as the primary deployment contract.
- Is provided solely for backward compatibility and debugging visibility.

## Key takeaways

- Publish first, then deploy (separation of structure and values).
- Artifacts are parameterized; resolution happens later.
- Integrations define actual publish/deploy behaviors.
- The system is extensible—build your own integration to target new platforms or internal tooling.

> - The legacy manifest can still be generated, but it's static and not evolving.

Reviewer: This would be better written with prose to improve readability. We don't always need lists.

The `aspire publish` command produces deployment artifacts that contain unresolved parameters (placeholders). The `aspire deploy` command uses these artifacts, resolves the parameters when supported by the target integration, and then executes the deployment. Some integrations don't support the `deploy` command.

## Aspire CLI commands (conceptual behavior)

| Command | What it does | Outputs | Parameter state | Requires integration capability |
|---------|--------------|---------|-----------------|---------------------------------|
| `aspire publish` | Transforms the application model into integration-specific assets (Compose files, manifests, specifications, etc.). | Intermediate artifacts (not directly production-final). | Unresolved (placeholders, e.g. `${VAR}` or similar). | Publish support |
| `aspire deploy` | Runs a deployment using one or more integrations (build, parameter resolution, apply). | Real resources / applied changes. | Resolved. | Deploy support |

> If an integration does not implement deploy functionality, `aspire deploy` will not deploy that target (it may warn or no-op for it).

Reviewer: The same is true for `aspire publish`. They both can either warn or no-op, depending on the corresponding hosting implementation. Most official Aspire integrations warn.

Reviewer: The text reads as a warning now, but you're correct, it "fails", for publish:

```Output
Step 1: Analyzing model.

       ✗ FAILED: Analyzing the distributed application model for publishing and deployment capabilities. 00:00:00
           No resources in the distributed application model support publishing.

❌ FAILED: Analyzing model. completed with errors
```

And for deploy:

```Output
Step 1: Analyzing model.

       ✗ FAILED: Analyzing the distributed application model for publishing and deployment capabilities. 00:00:00
           No resources in the distributed application model support deployment.

❌ FAILED: Analyzing model. completed with errors

════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════

✗ DEPLOYMENT FAILED: Deployment completed with errors
```

This should be reflected here more accurately. Also, the order should be publish then deploy, to align with the proper order of operations. When I first read this sentence, I thought it was going to miss that it also applies to publish.

When you run `aspire deploy` without any integrations that support deployment, you'll see this error:

```Output
FAILED: Analyzing the distributed application model for publishing and deployment capabilities.
           No resources in the distributed application model support deployment.
```

Similarly, when you run `aspire publish` without any integrations that support publishing, you'll see:

```Output
Analyzing the distributed application model for publishing and deployment capabilities. 00:00:00
           No resources in the distributed application model support publishing.

> ```

Reviewer: I'd change the order here to show `aspire `publish` first, since it occurs first logically in all scenarios (I'm actually curious is it should be considered a prerequisite to `aspire deploy` since you must call it first in all cases that I'm familiar with. I cannot think of a scenario where you wouldn't call `aspire publish`, but we might want to call that out too).

> These messages indicate that you need to add deployment integrations to your AppHost project. Deployment integrations are NuGet packages (like `Aspire.Hosting.Docker`, `Aspire.Hosting.Kubernetes`, or `Aspire.Hosting.Azure`) that provide the publishing and deployment capabilities for specific target platforms.

Reviewer: I'd prefer to not include a new integration category, these are simply hosting integrations. I'd shy away from calling them deployment integrations.

## Parameter placeholders

Published assets intentionally contain placeholders instead of concrete values. For Docker Compose–based publish output, parameterization appears as standard environment variable references. For example, a publish artifact might include:

```yaml
services:
  pg:

>     image: "docker.io/library/postgres:17.2"

Reviewer: ```suggestion
    image: "docker.io/library/postgres:17.6"
```

      ASPNETCORE_FORWARDEDHEADERS_ENABLED: "true"
      HTTP_PORTS: "8001"
      ConnectionStrings__db: "Host=pg;Port=5432;Username=postgres;Password=${PG_PASSWORD};Database=db"
    ports:
      - "8002:8001"
      - "8004:8003"
    depends_on:
      pg:
        condition: "service_started"
    networks:
      - "aspire"
```

Key points:

- `${PG_PASSWORD}` and `${DBSETUP_IMAGE}` (and similar) are *placeholders* in the published asset.
- They are not resolved during `aspire publish`.
- A deployment engine (which could be `aspire deploy`, `docker compose` plus a script exporting variables, CI/CD variable injection, etc.) supplies their values later.
- This keeps secrets and environment-specific values decoupled from generated structure.

Different integrations may use different placeholder conventions (environment variables, tokens, or parameter metadata), but the principle remains: publish preserves *shape*, deploy injects *values*.

## Publisher model and compute environments

.NET Aspire uses a flexible publisher model that distributes publishing behavior across your application graph. Resources support publishing and deployment through annotations:

> - `PublishingCallbackAnnotation` can be added to a resource to support `aspire publish`
> - `DeployingCallbackAnnotation` can be added to a resource to support `aspire deploy`

Reviewer: ```suggestion
- <xref:Aspire.Hosting.ApplicationModel.PublishingCallbackAnnotation> can be added to a resource to support `aspire publish`
- <xref:Aspire.Hosting.ApplicationModel.DeployingCallbackAnnotation> can be added to a resource to support `aspire deploy`
```

This design enables hybrid and heterogeneous deployments, where different services within the same app can be deployed to different targets (cloud, edge, local).

### Compute environments

A **compute environment** is a core deployment concept in .NET Aspire that represents a target platform where your application resources will be deployed. Compute environments define how resources should be transformed and what deployment artifacts should be generated.

**Compute resources** are the runnable parts of your application, such as .NET projects, containers, and executables that need to be deployed to a compute environment.

When you add a compute environment like Docker Compose or Kubernetes, Aspire applies the correct publishing behavior to all compatible compute resources in your app model—no extra configuration needed.

### Multiple environments require disambiguation

> If you add multiple compute environments, Aspire needs to know which resource goes where. Compute environments apply their transformations to all applicable compute resources (projects, containers, executables). If more than one environment matches a given resource, Aspire throws an ambiguous environment exception at publish time.

Reviewer: I tried finding this exception that is mentioned, but could not find it in the Aspire repo. I want to confirm it's accurate.

You can resolve this by using `WithComputeEnvironment(...)`:

```csharp
var k8s = builder.AddKubernetesEnvironment("k8s-env");
var compose = builder.AddDockerComposeEnvironment("docker-env");

builder.AddProject<Projects.Frontend>("frontend")
    .WithComputeEnvironment(k8s);

builder.AddProject<Projects.Backend>("backend")
    .WithComputeEnvironment(compose);
```

This example shows how you could explicitly map services to different compute environments. For example, a frontend in Kubernetes and a backend in Docker Compose.

## Hosting integration support matrix

| Integration Package | Target | Publish | Deploy | Notes |
|---------------------|--------|---------|--------|-------|

> | `Aspire.Hosting.Docker` | Docker / Docker Compose | Yes | No | Use generated Compose with your own scripts or tooling. |
> | `Aspire.Hosting.Kubernetes` | Kubernetes | Yes | No | Apply with `kubectl`, GitOps, or other controllers. |
> | `Aspire.Hosting.Azure.AppContainers` | Azure Container Apps | Yes | Yes (Preview) | Deploy capability is in Preview and may change. |
> | `Aspire.Hosting.Azure.AppService` | Azure App Service | Yes | Yes (Preview) | Deploy capability is in Preview and may change. |

Reviewer: ```suggestion
| `Aspire.Hosting.Docker` | Docker / Docker Compose | ✅ Yes | ❌ No | Use generated Compose with your own scripts or tooling. |
| `Aspire.Hosting.Kubernetes` | Kubernetes | ✅ Yes | ❌ No | Apply with `kubectl`, GitOps, or other controllers. |
| `Aspire.Hosting.Azure.AppContainers` | Azure Container Apps | ✅ Yes | ✅ Yes (Preview) | Deploy capability is in Preview and may change. |
| `Aspire.Hosting.Azure.AppService` | Azure App Service | ✅ Yes | ✅ Yes (Preview) | Deploy capability is in Preview and may change. |
```

**Compute environments:**

> - `AddDockerComposeEnvironment(...)` - Docker Compose deployment target
> - `AddKubernetesEnvironment(...)` - Kubernetes deployment target  
> - `AddAzureContainerAppEnvironment(...)` - Azure Container Apps deployment target
> - `AddAzureAppServiceEnvironment(...)` - Azure App Service deployment target

Reviewer: ```suggestion
- <xref:Aspire.Hosting.DockerComposeEnvironmentExtensions.AddDockerComposeEnvironment*> - Docker Compose deployment target.
- <xref:Aspire.Hosting.KubernetesEnvironmentExtensions.AddKubernetesEnvironment*> - Kubernetes deployment target .
- <xref:Aspire.Hosting.AzureContainerAppExtensions.AddAzureContainerAppEnvironment*> - Azure Container Apps deployment target.
- <xref:Aspire.Hosting.AzureAppServiceEnvironmentExtensions.AddAzureAppServiceEnvironment*> - Azure App Service deployment target.
```

title: .NET Aspire deployments
description: Learn about essential deployment concepts for .NET Aspire.
ms.topic: overview
ms.date: 09/26/2025
---

# .NET Aspire deployment overview

.NET Aspire separates the act of *producing deployment assets* from *executing a deployment*. The Aspire CLI (`aspire`) provides two high‑level entrypoints:

- `aspire publish` – Generates intermediate, parameterized assets for one or more hosting integrations.
- `aspire deploy` – Executes a deployment (when an integration implements deploy semantics) by resolving parameters and applying changes to a target environment.

> These commands are generic orchestration surfaces. Actual behavior (what gets generated, how deployment happens) comes from **hosting integrations** you reference (for example: Docker, Kubernetes, Azure). The system is **extensible**—you can build your own publishing or deployment integrations that plug into the same model.

Reviewer 2: They add resources that have deployment and publish annotations. They are added and used in similar ways.

> These commands are generic orchestration surfaces. Actual behavior (what gets generated, how deployment happens) comes from **hosting integrations** you reference (for example: Docker, Kubernetes, Azure). The system is **extensible**—you can build your own publishing or deployment integrations that plug into the same model.

Reviewer 2: Agree with this. I would change the wording here to be more direct. aspire deploy *does* execute the deployment and does not necessarily produce assests.

- `${PG_PASSWORD}` and `${DBSETUP_IMAGE}` (and similar) are *placeholders* in the published asset.
- They are not resolved during `aspire publish`.
- A deployment engine (which could be `aspire deploy`, `docker compose` plus a script exporting variables, CI/CD variable injection, etc.) supplies their values later.
- This keeps secrets and environment-specific values decoupled from generated structure.

Different integrations may use different placeholder conventions (environment variables, tokens, or parameter metadata), but the principle remains: publish preserves *shape*, deploy injects *values*.

## Publisher model and compute environments

.NET Aspire uses a flexible publisher model that distributes publishing behavior across your application graph. Resources support publishing and deployment through annotations:

- `PublishingCallbackAnnotation` can be added to a resource to support `aspire publish`
- `DeployingCallbackAnnotation` can be added to a resource to support `aspire deploy`

This design enables hybrid and heterogeneous deployments, where different services within the same app can be deployed to different targets (cloud, edge, local).

### Compute environments

> A **compute environment** is a core deployment concept in .NET Aspire that represents a target platform where your application resources will be deployed. Compute environments define how resources should be transformed and what deployment artifacts should be generated.

Reviewer 4: Should we define what a "compute resource" is?

Reviewer 2: Yep, it's sort of jammed in here.

Most apps only need one compute environment, such as:

```csharp
builder.AddAzureContainerAppEnvironment("env");
```

When you add a compute environment, Aspire applies the correct publishing behavior to all compatible compute resources in your app model—no extra configuration needed.

### Multiple environments require disambiguation

If you add multiple compute environments, Aspire needs to know which resource goes where. Compute environments apply their transformations to all applicable compute resources (projects, containers, executables). If more than one environment matches a given resource, Aspire throws an ambiguous environment exception at publish time.

You can resolve this by using `WithComputeEnvironment(...)`:

```csharp
var k8s = builder.AddKubernetesEnvironment("k8s-env");
var compose = builder.AddDockerComposeEnvironment("docker-env");

builder.AddProject<Projects.Frontend>("frontend")
    .WithComputeEnvironment(k8s);

builder.AddProject<Projects.Backend>("backend")
    .WithComputeEnvironment(compose);
```

> This example shows how you could explicitly map services to different compute targets—modeling, for example, a frontend in Kubernetes and a backend in Docker Compose.

Reviewer 4: ```suggestion
This example shows how you could explicitly map services to different compute environments. For example, a frontend in Kubernetes and a backend in Docker Compose.
```

Earlier workflows emphasized a single "deployment manifest" generated from specialized AppHost targets. The modern approach centers on `aspire publish` + integration extensibility. The legacy manifest format is **not being evolved further**, but you can still generate it for inspection or debugging:

```bash
aspire publish --publisher manifest -o diagnostics/
```

This:

- Produces a manifest snapshot useful for understanding resource graphs or troubleshooting.
- Should not be treated as the primary deployment contract.
- Is provided solely for backward compatibility and debugging visibility.

## Key takeaways

- Publish first, then deploy (separation of structure and values).
- Artifacts are parameterized; resolution happens later.
- Integrations define actual publish/deploy behaviors.
- The system is extensible—build your own integration to target new platforms or internal tooling.
- The legacy manifest can still be generated, but it's static and not evolving.

## See also

- Hosting integration reference packages (`Aspire.Hosting.*`)
- Observability and telemetry guidance
- Secret and parameter handling recommendations
- Azure deployment (preview) notes

> - Community tooling (Aspir8)

Reviewer 4: These should be links.

                new InteractionInput
                {
                    Label = "Environment Name",
                    InputType = InputType.Text,
                    Required = true,
                    Placeholder = "dev, staging, prod"
                },
                cancellationToken: ct);

            // Custom deployment logic here
            var reporter = ctx.ActivityReporter;
            await using (var deployStep = await reporter.CreateStepAsync(
                $"Deploying data seed job to {envResult.Value}", ct))
            {
                // Simulate deployment work
                await Task.Delay(2000, ct);
                await deployStep.SucceedAsync("Data seed job deployed successfully", ct);
            }
        }));

        return resourceBuilder;
    }
}
```

This custom deployment logic integrates seamlessly with the `aspire deploy` command, providing interactive prompts and progress reporting.

> [!NOTE]
> While the `DeployingCallbackAnnotation` API is available, there are currently no built-in resources that natively support deployment callbacks. Built-in resource support will be added in future versions.

> [!IMPORTANT]

> > The `aspire deploy` command requires enabling a feature flag: `aspire config set features.deployCommandEnabled true`

Reviewer 2: Undo this, this is not required. The feature flag is done.
